### PR TITLE
Adding '--cgroup-parent' option.

### PR DESCRIPTION
**Pull request by vishh**
#8551

The parent cgroup can be absolute or relative to the cgroup of the init process. Docker will create the parent cgroup if it does not exist.

Note: I am having trouble getting integration tests to work locally. Will update the PR once the tests have been validated.

cc @crosbymichael @LK4D4

### CODE REVIEW

**Comment by jfrazelle** [comment]
hmmm you have sigs, ignore gordon

**Comment by vishh** [comment]
@jfrazelle: I get the following error whenever I run integration tests.
```
+ go test -test.timeout=30m github.com/docker/docker/integration
time="2015-03-16T23:50:14Z" level=error msg="Failed to read /proc/sys/net/ipv4/ip_local_port_range kernel parameter: open /proc/sys/net/ipv4/ip_local_port_range: no such file or directory"
Pulling repository docker-test-image^M
Pulling image (latest) from docker-test-image ^M
Pulling image (latest) from docker-test-image, endpoint: https://registry-1.docker.io/v1/ ^M
Pulling dependent layers ^M
Pulling metadata ^M
Pulling fs layer ^M
Downloading [>
...
Error downloading dependent layers ^M
Error pulling image (latest) from docker-test-image, endpoint: https://registry-1.docker.io/v1/, unable to decode ApplyLayer JSON response: invalid character 'i' in literal true (expecting 'r') ^M
Error pulling image (latest) from docker-test-image, unable to decode ApplyLayer JSON response: invalid character 'i' in literal true (expecting 'r') ^M
Download complete ^M
time="2015-03-16T23:50:18Z" level=fatal msg="Unable to pull the test image: could not find image: no such id: 83599e29c455eb719f77d799bc7c51521b9551972f5a850d7ad265bc1b5292f6"
exit status 1
```
Any idea why I get this error?

**Comment by jfrazelle** [comment]
hmmm this could be because registry is being weird ping @dmcgowan idk... we need to get rid of the integration tests altogether

**Comment by vishh** [comment]
@jfrazelle: I see similar errors with the cli integration tests as well.

**Comment by vishh** [comment]
@jfrazelle: When is the code freeze for 1.6.0? Any chance that this feature will make it into 1.6.0? This feature will help improve node reliability quite a bit in Kubernetes.

**Comment by jfrazelle** [comment]
it would need to be in by sunday, we can try our best, it already had design review on the issue right?
ping @crosbymichael

**Comment by vishh** [comment]
Yup. AFAIU design review is done.

**Comment by LK4D4** [comment]
@vishh I think it has pretty good chance to go in 1.6 if we'll fix tests :)

**Comment by jfrazelle** [comment]
Can we change the title to not WIP if it's for potential inclusion in 1.6 :)

**Comment by vishh** [comment]
@LK4D4: The tests are now passing in linux. What is the recommended way to skip linux specific integration tests in windows?

**Comment by vishh** [comment]
@jfrazelle: Removed `WIP` from the title.

**Comment by jfrazelle** [comment]
i think you can add to _unix.go file in tests or there are skip testrequires functions

**Comment by jfrazelle** [comment]
but yayyyy! so close

**Comment by jfrazelle** [comment]
the implementation is so simple :) LGTM after windows is skipped for the tests

**Comment by vishh** [comment]
@jfrazelle: Thanks for the tip. I moved the tests to an equivalent "_unix" file.

**Comment by LK4D4** [comment]
Actually `_unix` postfix has no special meaning. `_linux` has.
ping @ahmetalpbalkan

**Comment by vishh** [comment]
@LK4D4: I am not sure what you meant by that, but the `*_unix_tets.go` files `build !windows` tag on them.

**Comment by LK4D4** [comment]
@vishh Oh, tag is okay. Just with `_linux` postfix you don't need tag.

**Comment by ahmetalpbalkan** [comment]
@vishh @LK4D4 convention we follow is mostly to have `_unix_test.go` on cli code and `_linux_test.go` on daemon code to prevent them from executed on windows. 

if this is a cgroup thing, file should be `_linux_test.go` and should have `// +build linux` rather than !windows.

@LK4D4 by convention, if the package will also be compiled on darwin/windows someday, we explicitly have the build tag declared on the source file... (another convention)

**Comment by LK4D4** [comment]
@ahmetalpbalkan Thanks! Maybe makes sense to write doc about that?

**Comment by vishh** [comment]
@ahmetalpbalkan: Thanks for the detailed response. 
@LK4D4: I am happy to rearrange the tests if required. Can we get this PR reviewed?

**Comment by LK4D4** [comment]
@vishh Of course, I'll review it now. But passing tests important too :)

**Comment by LK4D4** [comment]
@vishh Looking forward to docs review: you need API docs and couple of words in *NEW* section in API.

**Comment by vishh** [comment]
@LK4D4: Thanks for the review :) Tests are very important and I am happy to add more tests if you think it helps. I have modified a few docs already. Which file holds the *NEW* section for 1.6 release?

**Comment by LK4D4** [comment]
@vishh `docs/sources/reference/api/docker_remote_api.md`, there are some examples.

**Comment by jfrazelle** [comment]
also i thin your tests will need testRequires(nativeExecDriver) because this probably doesnt work on lxc

**Comment by vishh** [comment]
@LK4D4: Updated the remote api docs.
@jfrazelle: The tests have `testRequires(nativeExecDriver)` already.

**Comment by vishh** [comment]
@jfrazelle: The windows build failure is most likely a flake and not related to this PR. Can you restart the windows build?
Here is the failing test log:
```
--- FAIL: TestPullVerified (11.32s)
	docker_cli_pull_test.go:71: pulling a verified image failed. expected: The image you are pulling has been verified
		got: Pulling repository hello-world
		e45a5af57b00: Pulling image (latest) from hello-world
		e45a5af57b00: Pulling image (latest) from hello-world, endpoint: https://registry-1.docker.io/v1/
		e45a5af57b00: Pulling dependent layers
		511136ea3c5a: Download complete
		31cbccb51277: Download complete
		e45a5af57b00: Download complete
		e45a5af57b00: Download complete
		Status: Image is up to date for hello-world:latest
		, <nil>
```

**Comment by jfrazelle** [comment]
ah man ya that is having problems i think it should be fine

**Comment by jfrazelle** [comment]
ping @crosbymichael for review :D

File: docs/sources/reference/api/docker_remote_api_v1.18.md [hunk 1]
```diff
@@ -163,6 +163,7 @@ Create a container
                "Devices": [],
                "Ulimits": [{}],
                "LogConfig": { "Type": "json-file", Config: {} }
+	       "CgroupParent": ""
```
Comment (bgrant0607) [inline docs/sources/reference/api/docker_remote_api_v1.18.md]: Whitespace looks off.

File: docs/sources/reference/commandline/cli.md [hunk 2]
```diff
@@ -802,6 +802,7 @@ Creates a new container.
       -c, --cpu-shares=0         CPU shares (relative weight)
       --cap-add=[]               Add Linux capabilities
       --cap-drop=[]              Drop Linux capabilities
+      --cgroup-parent=           Optional parent cgroup for the container
```
Comment (bgrant0607) [inline docs/sources/reference/commandline/cli.md]: All other examples include the default value for the datatype, such as `""`.

File: docs/sources/reference/api/docker_remote_api_v1.18.md (hunk 1 above)
Comment (vishh) [inline docs/sources/reference/api/docker_remote_api_v1.18.md]: Fixed.

File: docs/sources/reference/commandline/cli.md (hunk 2 above)
Comment (vishh) [inline docs/sources/reference/commandline/cli.md]: Fixed.

**Comment by LK4D4** [comment]
@vishh Hmmm, stats test failing. Also your test failing too.

**Comment by vishh** [comment]
@LK4D4: The tests failed once on a systemd machine. Are the tests scheduled on arbitrary linux distros? I added some more debug logs to capture failures. Can you re-run the tests?

**Comment by vishh** [comment]
Ping @crosbymichael

**Comment by LK4D4** [comment]
LGTM

**Comment by crosbymichael** [comment]
LGTM

**Comment by jfrazelle** [comment]
omg yayyy!

**Comment by vishh** [comment]
Thanks for the quick review @LK4D4 and @crosbymichael. I hope the integration tests don't remain flaky. If they fail ping me and I will fix them.

**Comment by smarterclayton** [comment]
@mrunalp can you look at adding this to builds?

**Comment by mrunalp** [comment]
@smarterclayton Yes, it will quite useful for builds as well.

**Pull request merged**